File: src/ScalaSolutions/src/medium/ReverseInteger.scala
package medium

object ReverseInteger extends App {
  def reverse(x: Int): Int = {
    val s = if (x < 0) x.toString.tail else x.toString
    val d = s.reverse.toDouble * x.signum
    if (d.isValidInt) d.toInt else 0
  }
}

File: src/HaskellSolutions/src/Medium/ReverseInteger.hs
module Medium.ReverseInteger where

import           Data.Bool (bool)
import           Data.List (isPrefixOf)

reverseInteger :: Integer -> Integer
reverseInteger int
  | int < (-2 ^ 31) || int >= (2 ^ 31) = 0
  | otherwise = (reverseInt . show) int
  where
    reverseInt x =
      bool (read x) ((negate . read . reverse . tail) x) ("-" `isPrefixOf` x)
--
-- $> reverseInteger (-2^31)
